core/src/main/java/edu/wpi/grip/core/sources/NetworkValueSource.java

```diff
@@ -0,0 +1,199 @@
+            this.valueType = ValueType.BOOLEAN;
+        } else if (typeProperty.equals("NUMBER")) {
+            this.booleanOutputSocket = null;
+            this.numberOutputSocket = new OutputSocket<>(eventBus, numberOutputHint);
+            this.stringOutputSocket = null;
+            this.outputSocket = numberOutputSocket;
+            this.valueType = ValueType.NUMBER;
+        } else if (typeProperty.equals("STRING")) {
+            this.booleanOutputSocket = null;
+            this.numberOutputSocket = null;
+            this.stringOutputSocket = new OutputSocket<>(eventBus, stringOutputHint);
+            this.outputSocket = stringOutputSocket;
+            this.valueType = ValueType.STRING;
+        } else {
+            throw new IllegalArgumentException("NetworkValueSource does not support type " + typeProperty);
+        }
+
+        table.addTableListener(key, (source, key, value, isNew) -> {
+            eventBus.post(new SourceHasPendingUpdateEvent(this));
+        }, true);
+    }
+
+    @Override
+    public String getName() {
+        return this.name;
+    }
+
+    @Override
+    public OutputSocket[] createOutputSockets() {
+        return new OutputSocket[]{outputSocket};
+    }
+
+    @Override
+    protected boolean updateOutputSockets() {
+        if (valueType == ValueType.BOOLEAN) {
+            boolean newValue = table.getBoolean(key, false);
+            // We have a new value then we need to update the socket value
+            if (newValue != booleanPrevValue) {
+                booleanPrevValue = newValue;
+                // Update the value
+                booleanOutputSocket.setValue(newValue);
+                // We have updated output sockets
+                return true;
+            } else {
+                return false; // No output sockets were updated
+            }
+        } else if (valueType == ValueType.NUMBER) {
+            double newValue = table.getNumber(key, 0);
+            // We have a new value then we need to update the socket value
+            if (newValue != numberPrevValue) {
+                numberPrevValue = newValue;
+                // Update the value
+                numberOutputSocket.setValue(newValue);
+                // We have updated output sockets
+                return true;
+            } else {
+                return false; // No output sockets were updated
+            }
+        } else if (valueType == ValueType.STRING) {
+            String newValue = table.getString(key, "");
+            // We have a new value then we need to update the socket value
+            if (!newValue.equals(stringPrevValue)) {
+                stringPrevValue = newValue;
+                // Update the value
+                stringOutputSocket.setValue(newValue);
+                // We have updated output sockets
+                return true;
+            } else {
+                return false; // No output sockets were updated
+            }
```

Comment:
I was thinking of refactoring the type-specific handling to an internal class to make it a bit cleaner and get rid of the null assignments, but I don't know how much code can be deduplicated because fundamentally there are different types and thus different function calls required in this section in particular.